### PR TITLE
PR: Add runcell to spydercustomize

### PR DESCRIPTION
**Pull request by bcolsen**
### Pull Request Checklist

* [x] Based your PR on the latest version of the correct branch (master or 3.x)
* [x] Followed [PEP8](https://www.python.org/dev/peps/pep-0008/) for code style
* [x] Ensured your pull request hasn't eliminated unrelated blank lines/spaces,
      modified the ``spyder/defaults`` directory, or added new icons/assets
* [x] Wrote at least one-line docstrings for any new functions
* [ ] Added at least one unit test covering the changes, if at all possible
* [x] Described your changes and the motivation for them below
* [x] Noted what issue(s) this pull request resolves, creating one if needed

## Description of Changes

This PR makes a function that runs a cell instead of copying the cell contents into the current console. With this function can allow for correct traceback links when used with https://github.com/spyder-ide/spyder/pull/7310

### Implementation

The function reads a code section from a variable in the ipython namespace of the console. The function executes this code then deletes the variable. The function displays an error if it is run from the console. The variable could remain so the function can run the last cell, but it's not obvious that it won't reflect any subsequent changes to the cell block in the editor. The cell_mane is currently only used to show the user a simple history of what cells have been run

### Issue(s) Resolved

https://github.com/spyder-ide/spyder/issues/7113

### CODE REVIEW

**Comment by bcolsen** [comment]
@ccordoba12 There is some code duplication here for the UserModuleReloader. 

I'm not sure how it works but can I just break it out into it's own function like this?
```python
def run_umr():
    global __umr__
        if os.environ.get("SPY_UMR_ENABLED", "").lower() == "true":
            if __umr__ is None:
                namelist = os.environ.get("SPY_UMR_NAMELIST", None)
                if namelist is not None:
                    namelist = namelist.split(',')
                __umr__ = UserModuleReloader(namelist=namelist)
            else:
                verbose = os.environ.get("SPY_UMR_VERBOSE", "").lower() == "true"
                __umr__.run(verbose=verbose)
 ```

**Comment by bcolsen** [comment]
@ccordoba12 This is ready for you to have a look at.

File: spyder_kernels/customize/spydercustomize.py [hunk 1]
```diff
@@ -616,6 +616,18 @@ def _get_globals():
     ipython_shell = get_ipython()
     return ipython_shell.user_ns
 
+def run_umr():
+    """Runs the user module reloader"""
```
Comment (CAM-Gerlach) [inline spyder_kernels/customize/spydercustomize.py]: Missing period, also should be imperative: ``Run the user module reloader.``

File: spyder_kernels/customize/spydercustomize.py [hunk 2]
```diff
@@ -674,6 +677,49 @@ def runfile(filename, args=None, wdir=None, namespace=None, post_mortem=False):
 builtins.runfile = runfile
 
 
+def runcell(cell_name, filename):
+    """
+    Runs a code cell from an editor as a file.
```
Comment (CAM-Gerlach) [inline spyder_kernels/customize/spydercustomize.py]: Again, should be imperative—``Run a code cell...``.

File: spyder_kernels/customize/spydercustomize.py [hunk 3]
```diff
@@ -674,6 +677,49 @@ def runfile(filename, args=None, wdir=None, namespace=None, post_mortem=False):
...
+
+    Currently looks for code in in a ipython property called `cell_code`.
+    This property must be set by the editor prior to calling this function
```
Comment (CAM-Gerlach) [inline spyder_kernels/customize/spydercustomize.py]: Missing a period.

File: spyder_kernels/customize/spydercustomize.py [hunk 4]
```diff
@@ -674,6 +677,49 @@ def runfile(filename, args=None, wdir=None, namespace=None, post_mortem=False):
...
+        pass
+    run_umr()
+    namespace = _get_globals()
+    namespace['__file__'] = filename
+    ipython_shell = get_ipython()
+    try:
+        cell_code = ipython_shell.cell_code
+    except AttributeError:
+        _print("--Run Cell Error--\n"
+               "Please use only in editor")
```
Comment (CAM-Gerlach) [inline spyder_kernels/customize/spydercustomize.py]: For clarity, you could say something like ``Please use only through Spyder's Editor; shouldn't be called manually.``

File: spyder_kernels/customize/spydercustomize.py [hunk 5]
```diff
@@ -674,6 +677,49 @@ def runfile(filename, args=None, wdir=None, namespace=None, post_mortem=False):
...
+
+    The filename is needed to allow for proper traceback links.
+
+    The cell_name is used currently as a reference in the history log of which
+    cell was run with the command. It is currently unused in the function.
+
+    Parameters
+    ----------
+
+    cell_name: The name of the code cell (string)
```
Comment (CAM-Gerlach) [inline spyder_kernels/customize/spydercustomize.py]: What docstring convention is this? It doesn't like like Googledoc, Numpydoc, plain reST or javadoc/epytext/etc. I don't believe we've ever formally decided on one since our docstrings are rather skimpy, but those we do have are either Googledoc or Numpydoc so probably best to stick with one of those, or at the very least a common format.

Comment (bcolsen) [reply spyder_kernels/customize/spydercustomize.py]: Yeah, it's not anything.  I'm going to go for Numpydoc because I'm most familiar with that.

File: spyder_kernels/customize/spydercustomize.py [hunk 6]
```diff
@@ -671,9 +676,48 @@ def runfile(filename, args=None, wdir=None, namespace=None, post_mortem=False):
...
 
+
 builtins.runfile = runfile
 
 
+def runcell(cellname, filename):
+    """
+    Run a code cell from an editor as a file.
+
+    Currently looks for code in in a `ipython` property called `cell_code`.
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder_kernels/customize/spydercustomize.py]: "Code in *an* ipython..."

File: spyder_kernels/customize/spydercustomize.py [hunk 7]
```diff
@@ -671,9 +676,48 @@ def runfile(filename, args=None, wdir=None, namespace=None, post_mortem=False):
...
+    ipython_shell = get_ipython()
+    try:
+        cell_code = ipython_shell.cell_code
+    except AttributeError:
+        _print("--Run Cell Error--\n"
+               "Please use only through Spyder's Editor; "
+               "shouldn't be called manually from the console")
+        return
+    ipython_shell.run_cell(cell_code)
+    del ipython_shell.cell_code
```
[the remark this answers is not in the document]
Comment (bcolsen) [reply spyder_kernels/customize/spydercustomize.py]: >So you won't be able to rerun a previous cell ran with runcell? I think that would seem a bit limiting to our users...

This is unfortunate but I think this would be a 'ah ok' moment for most users and not a deal breaker, since users can't run a cell from the console right now. Cell based work is mostly all in the editor anyway (change code -> hotkey) it seems unlikely that a user would return to the console after making a change to a cell in the editor.

I guess the ideal interface is call `runcell('my_cell', 'my_file')` and have `runcell` execute the right code cell in the right file. But code cells without names or the same name end that idea. 

Another interface is just call `runcell()` in the console and the IPython kernel asks the editor for the code in the cell that the cursor is in, but I believe this isn't currently possible with qtconsole and `runcell()` still might not do what users expect

All these things considered it seems code cells are best run from the code cell itself and not the console. This mirrors the notebook experience.


>So for that to work (from the UX point of view) we probably need to avoid saving runcell in IPython's history.

I agree, it certainly shouldn't be in the "up button" for previously used commands.  It's not useful there at all.  I'll see what I can do, but the difficulty of changing this command history was what lead me to make `runcell`.

My original goal of all this was to have a filter on the history to not include code executed from code cells and `runfile` ( see https://github.com/spyder-ide/spyder/issues/7233) but there was no way to differentiate code cells from anything else because they where just copied code just like anything else.  I wanted this because pressing up and down through long blocks of code was pretty error prone and the history log wasn't overly useful.

At this point it only has a name and a file as the arguments because I thought it would be good have an indication to the user of what code it was running currently. Having this function call in the history log might help the user see what code was run at that point to give context to the surrounding commands

[the remark this answers is not in the document]
Comment (bcolsen) [reply spyder_kernels/customize/spydercustomize.py]: One more thing runfile also doesn't save my file when I run it from the console which would not be expected by the user.

TLDR;

I think that history filtering would be great and is best left for another PR because far as I remember I think we would have make changes to qtconsole to do it properly.

[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder_kernels/customize/spydercustomize.py]: Ok, I can help you with that, don't worry. I think I know how to do it without much trouble (probably from the Spyder side, not here).

For now, please fix the little grammar errors me and @CAM-Gerlach pointed out in one of your docstrings, then I'll merge this one.

Great work, by the way!! I really like your solution!

[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder_kernels/customize/spydercustomize.py]: Also, do you need help with a test? I can help you with that too.

[the remark this answers is not in the document]
Comment (bcolsen) [reply spyder_kernels/customize/spydercustomize.py]: >For now, please fix the little grammar errors me and @CAM-Gerlach pointed out in one of your docstrings, then I'll merge this one.

I'll fix this up tonight.

>Great work, by the way!! I really like your solution!

It work pretty well now. Lots of ideas from @jitseniesen in here too!

**Comment by bcolsen** [comment]
> Also, do you need help with a test? I can help you with that too.

That would be great. I guess I would need to start a new test file for `spydercustomize` and that might be a bit over my head.

**Comment by ccordoba12** [comment]
> That would be great. I guess I would need to start a new test file for spydercustomize and that might be a bit over my head.

Mmh, I'll open an issue for that once this is merged because I can add test very easily with the improvements I added in the `ipdb` branch, but not here.

**Comment by ccordoba12** [comment]
Ok, let's merge this so the Spyder side can work.
